key-manager/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A crate used to derive keys useful for the Oxide control plane
+
+use std::collections::BTreeMap;
+
+use async_trait::async_trait;
+use hkdf::Hkdf;
+use secrecy::{ExposeSecret, Secret};
+use sha3::Sha3_256;
+use sled_hardware::DiskIdentity;
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+/// Input Key Material
+///
+/// This should never be used directly, and always wrapped in a `Secret<Ikm>`
+/// upon construction. We sparate the two types, because a `Secret` must contain
+/// `Zeroizable` data, and a `Box<[u8; 32]>` is not zeroizable on its own.
+#[derive(Zeroize, ZeroizeOnDrop)]
+pub struct Ikm(pub Box<[u8; 32]>);
+
+/// Secret Input Key Material for a given rack reconfiguration epoch
+pub struct VersionedIkm {
+    pub epoch: u64,
+    pub salt: [u8; 32],
+    pub ikm: Secret<Ikm>,
+}
+
+impl VersionedIkm {
+    pub fn new(epoch: u64, salt: [u8; 32], data: &[u8; 32]) -> VersionedIkm {
+        let ikm = Secret::new(Ikm(Box::new(*data)));
+        VersionedIkm { epoch, salt, ikm }
+    }
+}
+
+/// An error returned by the [`KeyManager`]
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Secret not loaded for {epoch}")]
+    SecretNotLoaded { epoch: u64 },
+
+    #[error("Failed to retreive secret: {0}")]
+    SecretRetreival(#[from] SecretRetrieverError),
+}
+
+/// Derived Disk Encryption key
+#[derive(Zeroize, ZeroizeOnDrop, Default)]
+pub struct Aes256GcmDiskEncryptionKey(pub Box<[u8; 32]>);
+
+pub struct VersionedAes256GcmDiskEncryptionKey {
+    pub epoch: u64,
+    pub key: Secret<Aes256GcmDiskEncryptionKey>,
+}
+
+/// The main mechanism used to derive keys from a shared secret for the Oxide
+/// control plane.
+pub struct KeyManager<S: SecretRetriever> {
+    // A mechanism for retrieving input key material
+    secret_retriever: S,
+
+    /// Pseudo-Random-Keys (PRKs) wrapped in an `Hkdf` structure so that we can
+    /// create keys from them with `HKDF-Expand` (`prk.expand_multi_info`).
+    ///
+    /// In the common case, we will only have a single PRK for everything.
```

Comment:
Maybe note that the keys of this map are epochs?

Reply:
Will do

Reply:
Comment updated in [75d92cf](https://github.com/oxidecomputer/omicron/pull/2990/commits/75d92cf2095a16ce622234f5c06c16d232e4a712)